BacktraceScreen.c

```diff
@@ -0,0 +1,405 @@
+   assert(objectPathStart >= 0);
+
+   char* basename = frame->backtracePanel->process->procExe + frame->backtracePanel->process->procExeBasenameOffset;
+   char* object = line + objectPathStart;
+   if (!basename || !object) {
+      return;
+   }
+
+   char* lastSlash = strrchr(object, '/');
+   if (lastSlash) {
+      object = ++lastSlash;
+   }
+
+   char* objectTrimmed = String_trim(object);
+   if (String_eq(basename, objectTrimmed)) {
+      RichString_setAttrn(out, CRT_colors[PROCESS_BASENAME], object - line, strlen(objectTrimmed));
+   }
+
+   free(objectTrimmed);
+}
+
+static void BacktraceFrame_display(const Object* super, RichString* out) {
+   const BacktraceFrame* const frame = (const BacktraceFrame* const)super;
+   assert(frame);
+
+   if (frame->index == BACKTRACE_FRAME_HEADER_INDEX) {
+      BacktraceFrame_displayHeader(frame, out);
+      return;
+   }
+
+   const BacktracePanelPrintingHelper* const printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   char* functionName = frame->functionName;
+   int maxFunctionNameLength = printingHelper->maxFunctionNameLength;
+   if ((displayOptions & DEMANGLE_NAME_FUNCTION) == DEMANGLE_NAME_FUNCTION &&
+         printingHelper->maxDemangledFunctionNameLength > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFunctionNameLength;
+      if (frame->demangleFunctionName) {
+         functionName = frame->demangleFunctionName;
+      }
+   }
+
+   xAsprintf(&functionName, "%s+0x%lx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjectNameLength;
+   if ((displayOptions & SHOW_FULL_PATH_OBJECT) == SHOW_FULL_PATH_OBJECT) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjectPathLength;
+   }
+
+   char* line = NULL;
+   int objectPathStart = -1;
+   int len = xAsprintf(&line, "%-*d 0x%0*lx %-*s %n%-*s",
+                       (int)printingHelper->maxNumberFrameLength, frame->index,
+                       (int)printingHelper->maxAddressLength - (int)strlen("0x"), frame->address,
+                       (int)maxFunctionNameLength, functionName,
+                       &objectPathStart,
+                       (int)objectLength, objectDisplayed ? objectDisplayed : "-");
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   if (functionName) {
+      free(functionName);
+   }
+
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* const frame1 = (const BacktraceFrame* const)object1;
+   const BacktraceFrame* const frame2 = (const BacktraceFrame* const)object2;
```

Comment:
Some `const` can be left out IMO.

Reply:
There is no useless `const` here, I hope the compiler will do a better optimization with all `const` :smile:

Reply:
Usually we don't declare `const` on the pointers themselves, because the compiler is easy to detect whether a pointer address is changed or not, and the `const` would limit too much to the developers changing code comparing to its benefits (as optimization hints).

I mean to remove the `const` on the right side of the asterisks:

```c
   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
```

Reply:
I understand your point of view. I changed. Thanks.